gateway-ha/src/main/java/io/trino/gateway/ha/router/QueryCountBasedRouter.java

```diff
@@ -0,0 +1,120 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.trino.gateway.ha.router;
+
+import io.trino.gateway.ha.clustermonitor.ClusterStats;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.stream.Collectors;
+
+public class QueryCountBasedRouter
+            extends StochasticRoutingManager
+{
+    private static final Logger log = LoggerFactory.getLogger(QueryCountBasedRouter.class);
+    private ArrayList<ClusterStats> clusterStats;
+    private static final ReentrantReadWriteLock lock = new ReentrantReadWriteLock(true);
```

Comment:
The operations on `clusterStats` are read all and replace all.
`ReentrantReadWriteLock` can be replaced by `AtomicReference<List<ClusterStats>>`.

Reply:
Good catch on the read, thanks!
Both suggestions for the synchronization are good.
I might need to modify the flow a bit to use the record, with record I won't be able to modify the member attributes.

Reply:
- I am going back to the class for `ClusterStats` as the router needs to update it, it can be achieved by copying the records back and forth, but it won't be performant.
- I have moved the reading/sorting and updating of the cluster stats at the router level - inside `QueryCountBasedRouter`, for 2 reasons.

  1. The logic is specific to the router.

  2. The logic needs to read the stats, sort and the update the counters in the `ClusterStats` in a single transaction. This would avoid the race condition while processing simultaneous requests.